src/main/java/com/puppycrawl/tools/checkstyle/checks/sizes/RecordComponentNumberCheck.java

```diff
@@ -0,0 +1,241 @@
+import com.puppycrawl.tools.checkstyle.utils.TokenUtil;
+
+/**
+ * <p>
+ * Checks the number of record components in the
+ * <a href="http://cr.openjdk.java.net/~gbierman/jep359/jep359-20191031/specs/records-jls.html">
+ * header</a> of a record definition.
+ * </p>
+ * <ul>
+ * <li>
+ * Property {@code max} - Specify the maximum number of components allowed in the header of a
+ * record definition.
+ * Type is {@code int}.
+ * Default value is {@code 8}.
+ * </li>
+ * <li>
+ * Property {@code accessModifiers} - Access modifiers of record definitions where
+ * the number of record components should be checked.
+ * Type is {@code com.puppycrawl.tools.checkstyle.checks.naming.AccessModifierOption[]}.
+ * Default value is {@code public, protected, package, private}.
+ * </li>
+ * </ul>
+ * <p>
+ * To configure the check:
+ * </p>
+ * <pre>
+ * &lt;module name="RecordComponentNumber"/&gt;
+ * </pre>
+ * <p>
+ * Java code example:
+ * </p>
+ * <pre>
+ * public record MyRecord1(int x, int y) { // ok, 2 components
+ *     ...
+ * }
+ *
+ * record MyRecord2(int x, int y, String str,
+ *                           Node node, Order order, Data data
+ *                           String location, Date date, Image image) { // violation, 9 components
+ *     ...
+ * }
+ * </pre>
+ * <p>
+ * To configure the check to allow 5 record components at all access modifier levels
+ * for record definitions:
+ * </p>
+ * <pre>
+ * &lt;module name="RecordComponentNumber"&gt;
+ *   &lt;property name="max" value="5"/&gt;
+ * &lt;/module&gt;
+ * </pre>
+ * <p>
+ * Java code example:
+ * </p>
+ * <pre>
+ * public record MyRecord1(int x, int y, String str) { // ok, 3 components
+ *     ...
+ * }
+ *
+ * public record MyRecord2(int x, int y, String str,
+ *                           Node node, Order order, Data data) { // violation, 6 components
+ *     ...
+ * }
+ * </pre>
+ * <p>
+ * To configure the check to allow 10 record components for a public record definition,
+ * but 3 for private record definitions:
+ * </p>
+ * <pre>
+ * &lt;module name="RecordComponentNumber"&gt;
+ *   &lt;property name="max" value="3"/&gt;
+ *   &lt;property name="accessModifiers" value="private"/&gt;
+ * &lt;/module&gt;
+ * &lt;module name="RecordComponentNumber"&gt;
+ *   &lt;property name="max" value="10"/&gt;
+ *   &lt;property name="accessModifiers" value="public"/&gt;
+ * &lt;/module&gt;
+ * </pre>
+ * <p>
+ * Java code example:
+ * </p>
+ * <pre>
+ * public record MyRecord1(int x, int y, String str) { // ok, public record definition allowed 10
+ *     ...
+ * }
+ *
+ * private record MyRecord2(int x, int y, String str, Node node) { // violation
+ *     ...                                // private record definition allowed 3 components
+ * }
+ * </pre>
+ * <p>
+ * Parent is {@code com.puppycrawl.tools.checkstyle.TreeWalker}
+ * </p>
+ * <p>
+ * Violation Message Keys:
+ * </p>
+ * <ul>
+ * <li>
+ * {@code too.many.components}
+ * </li>
+ * </ul>
+ *
+ * @since 8.36
+ */
+@FileStatefulCheck
```

Comment:
it is stateless.

stateless means ready for  Multithread mode (our future)

Reply:
Is this in reference to the AtomicInteger usage?

Reply:
Fixed.